composer/datasets/synthetic.py

```diff
@@ -21,6 +36,201 @@ class SyntheticDataLabelType(StringEnum):
+    try:
+        from tokenizers import Tokenizer
+        from transformers import PreTrainedTokenizer
+    except ImportError as e:
+        raise ImportError(
+            textwrap.dedent("""\
+            Composer was installed without NLP support. To use NLP with Composer, run `pip install mosaicml[nlp]`
+            if using pip or `conda install -c conda-forge transformers tokenizers` if using Anaconda.""")) from e
+
+    # generate a synthetic dataset with reasonable defaults is none is provided
+    if dataset is None:
+        num_samples = 100
+        chars_per_sample = 128
+        column_names = ['sentence']
+        dataset = SyntheticHFDataset(num_samples=num_samples,
+                                     chars_per_sample=chars_per_sample,
+                                     column_names=column_names).generate_dataset()
+
+    # flatten dataset to get a simple stream of text for the tokenizer
+    dataset = [dataset[key] for key in dataset.column_names if key != 'idx']
+    dataset = [i for sublist in dataset for i in sublist]
+
+    if "bert" in tokenizer_family:
+        tokenizer_params = generate_bert_tokenizer_params(dataset)
+    elif "gpt2" in tokenizer_family:
+        tokenizer_params = generate_gpt2_tokenizer_params()
+    else:
+        raise ValueError(f"Synthetic tokenizers for tokenizer family {tokenizer_family} are currently unsupported.")
+
+    tokenizer = Tokenizer(tokenizer_params.tokenizer_model)
+    tokenizer.enable_padding(direction="right",
+                             pad_id=0,
+                             pad_type_id=0,
+                             pad_token=tokenizer_params.pad_token,
+                             pad_to_multiple_of=8)
+    tokenizer.normalizer = tokenizer_params.normalizer  # type: ignore
+    tokenizer.pre_tokenizer = tokenizer_params.pre_tokenizer  # type: ignore
+    tokenizer.decoder = tokenizer_params.decoder  # type: ignore
+    tokenizer_trainer = tokenizer_params.trainer_cls(
+        vocab_size=vocab_size,
+        initial_alphabet=tokenizer_params.initial_alphabet,
+        special_tokens=tokenizer_params.special_tokens,
+    )
+    tokenizer.train_from_iterator(dataset, trainer=tokenizer_trainer)
+
+    # save the tokenizer config
+    tmp_tokenizer_dir = mkdtemp()
+    tmp_tokenizer_file = join(tmp_tokenizer_dir, "tokenizer.json")
+    tokenizer.save(tmp_tokenizer_file)
+
+    # save the vocabulary and potential merges file
+    tokenizer_params.tokenizer_model.save(tmp_tokenizer_dir)  # type: ignore
+
+    # the .from_pretrained method doesn't load our padding for some reason, so we save it as a special kwarg
+    tmp_tokenizer_config = join(tmp_tokenizer_dir, "tokenizer_config.json")
+    with open(tmp_tokenizer_config, "w") as f:
+        json.dump({"pad_token": tokenizer_params.pad_token}, f)
+
+    # instantiate the new tokenizer
+    assert issubclass(tokenizer_params.tokenizer_cls, PreTrainedTokenizer)
```

Comment:
is this validation or an invariant?

Reply:
Validation! I just changed the assert into a ValueError!

Reply:
...wow, i sound way too excited about that